pbxd/pbx/definity.py

```diff
@@ -0,0 +1,411 @@
+        elif index == 1:  # EOF
+            self.logger.error('Connection failed with EOF at command prompt verify:\n{}'.format(self.session.before))
+            raise Exception('Connection failed with EOF at command prompt verify:\n{}'.format(self.session.before))
+
+        self.connected_termtype = termtype
+
+    def _ossi_object(self, response_fields, response_data):
+        """
+        Convert the OSSI field and data lists to a dictionary.
+        """
+        ossi_obj = {}
+        for i, field in enumerate(response_fields):
+            ossi_obj[field] = response_data[i]
+        if len(response_fields) != len(ossi_obj):
+            # there have been cases of duplicate field ids in some commands
+            self.logger.error('duplicate field ids detected {} != {}'.format(response_fields, ossi_obj.keys()))
+        return ossi_obj
+
+    def ossi_command(self, command, fields=None, debug=False):
+        """
+        Send a command to the PBX and return the result.
+
+        Provide a dictionary of field ids and values to change fields or return
+        specific fields.
+
+        One way to identify field ids is to run a "display" command and then
+        compare it to the output of the same command in a VT220 terminal.
+
+        Note: there have been cases of duplicate field ids from the PBX so the
+        data_list is available without ids if needed.
+        """
+        # switch back to the original ossi OSSI terminal type
+        self._select_termtype('ossi4')
+
+        # Send OSSI command
+        self.logger.info('command: {}'.format(command))
+        self.session.sendline('c {}'.format(command))  # command
+
+        # if no fields are specificed the pbx returns all fields
+        if fields is not None and len(fields) > 0:
+            self.logger.debug('fields: {}'.format(fields))
+            ids = ('\t'.join(sorted(fields))).strip()
+            data = ('\t'.join([fields[k] for k in sorted(fields)])).strip()
+            self.logger.debug('send: f{}'.format(ids))
+            self.session.sendline('f{}'.format(ids))  # fields
+            self.logger.debug('send: d{}'.format(data))
+            self.session.sendline('d{}'.format(data))  # data
+
+        self.session.sendline('t')  # command terminator
+
+        # Read the response and collect the OSSI objects
+        response_fields = []
+        response_data = []
+        response_error = None
+        complete_output = False
+        ossi_objects = []
+        raw_lines = []
+        while not complete_output:
+            index = self.session.expect(
+                [
+                    pexpect.TIMEOUT,
+                    pexpect.EOF,
+                    r'f[\S\t]+[\r\n]+',
+                    r'd[\S\t ]*[\r\n]+',
+                    r'e[\S\t ]+[\r\n]+',
+                    r'n[\r\n]+',
+                    r't[\r\n]+',
+                    r'c [\S ]+[\r\n]+',
```

Comment:
Concisely document these line formats nearby or link to authoritative doc.

Reply:
There is a comment with each match handler that describes what the line contains.
I've never read any official documentation from the vendor, although it was rumored to have existed at one point.

Reply:
Those comments are kinda too concise for my liking. Since this is apparently an interface deliberately undocumented by the vendor it would be useful to have our own more complete documentation of it.

Reply:
I'll try to write something more detailed here.